src/datasources/prices-api/prices-api.service.ts

```diff
@@ -0,0 +1,118 @@
+    this.defaultExpirationTimeInSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.default',
+      );
+    this.pricesTtlSeconds = this.configurationService.getOrThrow<number>(
+      'prices.pricesTtlSeconds',
+    );
+    this.defaultNotFoundExpirationTimeSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.notFound.default',
+      );
+  }
+
+  async getNativeCoinPrice(args: {
+    nativeCoinId: string;
+    fiatCode: string;
+  }): Promise<AssetPrice> {
+    try {
+      const cacheDir = CacheRouter.getNativeCoinPriceCacheDir(args);
+      const url = `${this.baseUrl}/simple/price?ids=${args.nativeCoinId}&vs_currencies=${args.fiatCode}`;
+      return await this.dataSource.get({
+        cacheDir,
+        url,
+        networkRequest: this._buildNetworkRequest(),
+        notFoundExpireTimeSeconds: this.defaultNotFoundExpirationTimeSeconds,
+        expireTimeSeconds: this.pricesTtlSeconds,
+      });
+    } catch (error) {
+      throw new DataSourceError(
+        `Error getting ${
+          args.nativeCoinId
+        } price from provider${this._mapProviderError(error)}`,
+      );
+    }
+  }
+  async getTokenPrice(args: {
+    chainName: string;
+    tokenAddress: string;
+    fiatCode: string;
+  }): Promise<AssetPrice> {
+    try {
+      const cacheDir = CacheRouter.getTokenPriceCacheDir(args);
+      const url = `${this.baseUrl}/simple/token_price/${args.chainName}?contract_addresses=${args.tokenAddress}&vs_currencies=${args.fiatCode}`;
+      const result: AssetPrice = await this.dataSource.get({
+        cacheDir,
+        url,
+        networkRequest: this._buildNetworkRequest(),
+        notFoundExpireTimeSeconds: this.defaultNotFoundExpirationTimeSeconds,
+        expireTimeSeconds: this.pricesTtlSeconds,
+      });
+      return result;
+    } catch (error) {
+      throw new DataSourceError(
+        `Error getting ${
+          args.tokenAddress
+        } price from provider${this._mapProviderError(error)}`,
+      );
+    }
+  }
+
+  async getFiatCodes(): Promise<string[]> {
+    try {
+      const cacheDir = CacheRouter.getPriceFiatCodesCacheDir();
+      const url = `${this.baseUrl}/simple/supported_vs_currencies`;
+      const result: string[] = await this.dataSource.get({
+        cacheDir,
+        url,
+        networkRequest: this._buildNetworkRequest(),
+        notFoundExpireTimeSeconds: this.defaultNotFoundExpirationTimeSeconds,
+        expireTimeSeconds: this.defaultExpirationTimeInSeconds,
+      });
+      return result;
+    } catch (error) {
+      throw new DataSourceError(
+        `Error getting Fiat Codes from prices provider${this._mapProviderError(
+          error,
+        )}`,
+      );
+    }
+  }
+
+  private _mapProviderError(error: any): string {
+    const errorCode = error?.status?.error_code;
+    return errorCode ? ` [status: ${errorCode}]` : '';
```

Comment:
What is the main reason for mapping the error code like this? The `DataSourceError` also has a `code` property which is not being used. Have you considered setting it also?

Reply:
Initially I thought `error_code` was only available in the response payload object:
```
{
    "status": {
        "error_code": 429,
        "error_message": "You've exceeded the Rate Limit. Please visit https://www.coingecko.com/en/api/pricing to subscribe to our API plans for higher rate limits."
    }
}
```

But it seems the status code is available as well, so I'm going to try to get it from there, thanks 👍

Reply:
It's also fine if the error code doesn't make sense to be set in the `DataSourceError`. The main goal of this error is to abstract away all the `datasource` related errors. If this error is not caught in the service then the `DataSourceErrorFilter` will return the following to the clients:

```javascript
{
  "code": <int>,
  "message": <string>
}
```

It's important to be aware that we should not leak the internals on why the request failed – in this case the Rate Limit exceeded on the CGW side should not be returned to the clients.